### PR TITLE
Binary search algo

### PR DESCRIPTION
**Pull request by sohbaker**
Adds updates to the search algorithm tasks and implements solution with tests for the Binary Search algorithm in Java. 

Will need to add tests, new `search` folder and interface to the student-accessible repo ([Intro to Java Course](https://github.com/cbfacademy/intro-to-java-course)).

### CODE REVIEW

**Review by Genyus — CHANGES_REQUESTED** [review]
Your [last commit](https://github.com/cbfacademy-solutions/java-course-exercises/pull/7/commits/b2a7366dbead4c32c8630e9b55f48f24044d1931) mixes test and solution code changes, so can't be cherry-picked into upstream. You can execute `git revert b2a7366d && git restore -s b2a7366d exercises/algorithms` to rollback the commit and restore the changes; then create two separate commits for the upstream and solution-only targets

**Comment by sohbaker** [comment]
Ah yes, thanks for catching that @Genyus.

File: exercises/algorithms/README.md [hunk 1]
```diff
@@ -81,35 +81,79 @@ return pIndex + 1
 
 ## :pushpin: Searching
 
-### Linear Search
```
Comment (tanyapowell) [inline exercises/algorithms/README.md:84]: Are we no longer teaching Linear Search?

Comment (sohbaker) [reply exercises/algorithms/README.md:84]: We're still teaching it but removed the exercise related to that content. Instead of trying to implement four algorithms, students will implement one search and one sort algorithm.

File: exercises/algorithms/src/main/java/com/cbfacademy/App.java [hunk 2]
```diff
@@ -1,7 +1,13 @@
...
+import com.cbfacademy.search.BinarySearch;
+
 public class App {
     public static void main(String[] args) {
-        System.out.println("Hello World!");
+        BinarySearch binarySearch = new BinarySearch();
+        int[] sortedArray = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
+        int target = 12;
+
+        System.out.println(binarySearch.search(sortedArray, target));
```
Comment (Genyus) [inline exercises/algorithms/src/main/java/com/cbfacademy/App.java]: Do we not want any examples of a successful search?

File: exercises/algorithms/src/main/java/com/cbfacademy/search/Search.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+     * @param sortedArray The array to be searched (the array should already be s
+     *                    rted)
+     * @param target      The value to be searched for in the array
+     */
+    int search(int[] sortedArray, int target);
+
+    /**
+     * Returns the exercise name
+     */
+    String getName();
```
Comment (Genyus) [inline exercises/algorithms/src/main/java/com/cbfacademy/search/Search.java:16]: Why is this needed? Seems out of scope for a search interface?

File: exercises/algorithms/src/test/java/com/cbfacademy/BinarySearchTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        assertTrue(result == 47);
+    }
+
+    @Test
+    @DisplayName("search returns the index of the target value when it is in the lower end of the array")
+    public void returnsIndexWhenTargetIsLowerEndOfArray() {
+        int result = binarySearch.search(testArray, 6);
+
+        assertTrue(result == 5);
+    }
```
Comment (Genyus) [inline exercises/algorithms/src/test/java/com/cbfacademy/BinarySearchTest.java]: These can be combined into a single parameterised test, e.g.

```java
    @ParameterizedTest
    @MethodSource
    @DisplayName("search() method returns index of target value")
    public void search_ReturnsIndexOfTargetValue(int target) {
        int index = binarySearch.search(testArray, target);

        assertTrue(index == target - 1);
    }

    private static search_ReturnsIndexOfTargetValue() {
        int maxIndex = testArray.length - 1;
        int[] targetIndices = { 0, 1, maxIndex / 4, maxIndex / 3, maxIndex / 2, maxIndex * 2 / 3, maxIndex * 3 / 4, maxIndex - 1, maxIndex}; // Test for values at relative positions rather than hard-coded indices
        List<Arguments> searchArguments = new ArrayList<>();

        for (int index : targetIndices) {
            searchArguments.add(arguments(testArray[index]));
        }

        return searchArguments.stream();
    }
```

File: exercises/algorithms/src/test/java/com/cbfacademy/BinarySearchTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    @DisplayName("returns this unit's name")
+    public void returnUnitName() {
+        String name = binarySearch.getName();
+
+        assertThat(name, is("Binary Search Algorithm Exercise"));
+    }
+
+    @Test
+    @DisplayName("search returns the index of the target value when it is in the middle of the array")
+    public void returnsIndexWhenTargetIsMidPointInArray() {
```
Comment (Genyus) [inline exercises/algorithms/src/test/java/com/cbfacademy/BinarySearchTest.java]: Just as a suggestion, prefixing the test name with the method name under test makes it easier to group and find tests for a specific method, e.g.

```java
public void search_ReturnsIndexWhenTargetIsMidPointInArray()
```

File: exercises/algorithms/src/test/java/com/cbfacademy/BinarySearchTest.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+import com.cbfacademy.search.BinarySearch;
+import com.cbfacademy.search.Search;
+
+@DisplayName(value = "BinarySearch Exercise")
+public class BinarySearchTest {
+    Search binarySearch;
+    int[] testArray = new int[50];
```
Comment (Genyus) [inline exercises/algorithms/src/test/java/com/cbfacademy/BinarySearchTest.java]: Test array can be made several magnitudes larger to better approximate real-world usage

**Review by Genyus — CHANGES_REQUESTED** [review]

File: exercises/algorithms/src/main/java/com/cbfacademy/App.java
Comment (Genyus) [inline exercises/algorithms/src/main/java/com/cbfacademy/App.java:1]: Following [Tanya's comment](https://github.com/cbfacademy-solutions/java-course-exercises/pull/9#discussion_r1325799205), might be a good idea if we move App.java into the respective search and sort folders

**Review by Genyus — CHANGES_REQUESTED** [review]

File: exercises/algorithms/src/main/java/com/cbfacademy/App.java
Comment (tanyapowell) [reply exercises/algorithms/src/main/java/com/cbfacademy/App.java:1]: Will you be doing the same for [the sort algorithm PR](#9)?

Comment (Genyus) [reply exercises/algorithms/src/main/java/com/cbfacademy/App.java:1]: Yep

File: exercises/algorithms/src/main/java/com/cbfacademy/search/Search.java (hunk 3 above)
Comment (sohbaker) [reply exercises/algorithms/src/main/java/com/cbfacademy/search/Search.java:16]: Heya, this follows on from my mention this week that it might be worth us thinking about standardising our exercise files (e.g. comments to describe a method, etc). We have previous exercises with a getName method and it's included here for consistency, though can be removed.

File: exercises/algorithms/src/main/java/com/cbfacademy/App.java (hunk 2 above)
Comment (sohbaker) [reply exercises/algorithms/src/main/java/com/cbfacademy/App.java]: Yes. Thanks for catching.

**Comment by sohbaker** [comment]
Raising a new PR

File: exercises/algorithms/src/main/java/com/cbfacademy/search/Search.java (hunk 3 above)
Comment (Genyus) [reply exercises/algorithms/src/main/java/com/cbfacademy/search/Search.java:16]: Yes, I've seen that in other exercises, but never understood the intended benefit for the students or assessors.

Very much in favour of standardisation, but let's aim for useful standards. 🙂

Comment (sohbaker) [reply exercises/algorithms/src/main/java/com/cbfacademy/search/Search.java:16]: I actually thought that the getName() on a few of the other tests was something you'd added. 

I see your point and happy to remove.
